### ISSUE TITLE
What does it do with symlinks?

### ISSUE THREAD

**Issue by malgorithms** [lgessler/kbrpc#2]
It's worth noting that KBFS allows symlinks...the idea in theory that you trust a partner, you trust to follow their symlink. However, we're still pondering this issue.

If I put a file in /keybase/private/me,you/ I can actually make it a symlink to something totally outside the folder, like some path on your machine. when you read it, you'll see your own private data. There could be some attacks with this. (In the keybase.pub case, for example, we don't want to be tricked into serving the server's own private files.)

Anyway, bringing this up because I haven't thought of what a malicious chat partner could do over your RPC layer by making their file link elsewhere. It's worth thinking about

**Comment by lgessler** [comment]
edit: as @atondwal pointed out to me, maybe I'm overthinking this, as anything that can be done in these ways could have also been done by just writing directly to the file. The only safeguard against that is detecting a malicious write before you do anything with it--that sounds pretty hard.

Yikes. So let's see. An attacker could either replace the `/keybase/private/me,you/private/.kbrpc` folder entirely or replace one of its files, e.g. `chat.lgessler.macbook.sent`[1], and replace it with arbitrary data.

There are two cases:

#### Symlink points to something on local machine not controlled by attacker

It's unclear what the risks are here--our file format is pretty idiosyncratic:

```
1459223606081lgesslerbGZrd2U=
1459224756550lgesslerbGt3amVmbGVrd2ZqbGZ3amVmd2VrZg==
1459224757421lgesslerZWZud2VsZg==
1459224760781lgesslercWx3amVxd2w=
```

(Note: the question mark character is ASCII 0x7 'bell', and the data is base 64 encoded--quick hacks we relied on to avoid thinking about escaping things, and a nice bonus about the encoding is that the file remains human readable. Hopefully 0x7 isn't allowable in a username, right?)

The attacker would need to turn KBRPC's expectation about file structure against it. It's not immediately clear to me how you would do evil things with this, as our format is (for better or worse) pretty idiosyncratic.

#### Symlink points to a something controlled by attacker

This is Pandora's box. Suppose the attacker put something in their public directory, e.g. `/keybase/public/attacker/cryptolocker-binary-in-kbrpc-data`. Suppose then the attacker symlinks to that from `/keybase/private/me,attacker/.kbrpc/chat.attacker.microsoftbob.sent` 

If the app is restricted just to printing the data maybe nothing will go so wrong, but as soon as you start allowing the execution of arbitrary code (in overt or subtle ways, like on a PDF) everything would go wrong.

As @emumoo pointed out in a discussion we had, KBRPC could just check everything it's working with and refuse to work with symlinks, but it sounds like the KBFS team is encouraging honoring symlinks if possible.

[1] `chat` is an additional layer of namespacing we added--probably should have been a directory instead, going to change that.

**Comment by atondwal** [comment]
Yeah, so as far as I can tell, all the attacker can do with symlinks is serve whatever data (which they could've done just by writing to the file) or make you append to a file that you don't want to: say link `/keybase/private/me,attacker/.kbrpc/chat.me.laptop.sent -> /home/me/.zshrc`

The latter seems like it could a good reason to disallow symlinks... OR we could just move `/keybase/private/me,u1,u2/.kbrpc/chat.me.laptop.sent` to say `/keybase/private/me#u1,u2/.kbrpc/chat.me.laptop.sent` so that only I can write to the files that I later write to.

...but I don't think there's any coherent way to stop someone sending you bad data signed by themselves... I suppose we could stop trusting the username metadata provided, and just get it from where we read the file, if we implement the above fix, but beyond that it's just data and a time. I suppose we should just tell clients that that metadata is not to be trusted to not have been tampered with by the person sending you the message (or have them roll their own)

**Comment by lgessler** [comment]
as @atondwal says, adopting a directory structure like `/keybase/private/me#u1,u2/.kbrpc/chat.me.laptop.sent` would remove attacks that rely on you writing to an unintended file or impersonating someone else. That leaves the domain of possible attacks (at least ones that are on our radar right now) within the files in `/keybase/private/attacker#me/.kbrpc/...`. 

That still allows the attacker to send whatever they want, but perhaps this is where we'd tell anyone building on top of this to add their own security layer.

**Comment by malgorithms** [comment]
We're internally talking a lot about a policy with symlinks in KBFS, so it's nice to see you all talking about these issues here. The short-term answer is that you should probably just check the files and insist they're not symlinks before using them.

Even this solution:

```
/keybase/private/a#b,c/foo
```

allows `a` (a bad guy) to make that a symbolic link to:

```
/keybase/private/a#b,c/foo -> /keybase/private/x#b,c/foo
```

So when `b` an `c` think they're getting a message from a, they're actually getting data that `x` wrote to them. This can confuse `b` and `c` and make them think `a` has stolen private data, or trip them up into replying with something they shouldn't.

This user is interesting to look at - a test user I made for keybase.pub unit tests...(enyoc is my last name backwards). it has symlinks to all kinds of different places:

https://keybase.pub/enyoc/keybase.pub-tests/symlinks/

The site recognizes that they're symlinks and won't serve any of that data raw. In the case where they're symlinks staying inside public folders, it links directly to the appropriate page but doesn't let you surf inside keybase.

For example, this actually works in KBFS:

```
# this prints a public file in MAX's folder, not enyoc's
cat /keybase/public/enyoc/keybase.pub-tests/symlinks/max_krohn_keys
```

yet keybase.pub won't honor it:

https://keybase.pub/enyoc/keybase.pub-tests/symlinks/max_krohn_keys
